jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DosHandler.java

```diff
@@ -0,0 +1,559 @@
+        installBean(_rejectHandler);
+    }
+
+    @Override
+    public void setServer(Server server)
+    {
+        super.setServer(server);
+        if (_rejectHandler instanceof Handler handler)
+            handler.setServer(server);
+    }
+
+    @Override
+    protected boolean onConditionsMet(Request request, Response response, Callback callback) throws Exception
+    {
+        // Reject if we have too many Trackers
+        if (_maxTrackers > 0 && _trackers.size() >= _maxTrackers)
+        {
+            // Try shrinking the tracker pool
+            long now = NanoTime.now();
+            _trackers.values().removeIf(tracker -> tracker.isIdle(now));
+            if (_trackers.size() >= _maxTrackers)
+                return _rejectHandler.handle(request, response, callback);
+        }
+
+        // Calculate an id for the request (which may be global empty string)
+        String id;
+        id = _getId.apply(request);
+
+        if (id == null)
+            return _rejectHandler.handle(request, response, callback);
+
+        // Obtain a tracker, creating a new one if necessary.  Trackers are removed if CyclicTimeouts#onExpired returns true
+        Tracker tracker = _trackers.computeIfAbsent(id, this::newTracker);
+
+        // If we are not over-limit then handle normally
+        if (!tracker.isRateExceeded(request.getBeginNanoTime(), true))
+            return nextHandler(request, response, callback);
+
+        // Otherwise reject the request
+        return _rejectHandler.handle(request, response, callback);
+    }
+
+    Tracker newTracker(String id)
+    {
+        return new Tracker(id, _rateControlFactory.newRate());
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        _cyclicTimeouts = new CyclicTimeouts<>(getServer().getScheduler())
+        {
+            @Override
+            protected Iterator<Tracker> iterator()
+            {
+                return _trackers.values().iterator();
+            }
+
+            @Override
+            protected boolean onExpired(Tracker tracker)
+            {
+                return tracker.isIdle(NanoTime.now());
+            }
+        };
+        super.doStart();
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        super.doStop();
+        _cyclicTimeouts.destroy();
+        _cyclicTimeouts = null;
+    }
+
+    /**
+     * A RateTracker is associated with a connection, and stores request rate data.
+     */
+    class Tracker implements CyclicTimeouts.Expirable
+    {
+        private final AutoLock _lock = new AutoLock();
+        private final String _id;
+        private final RateControl _rateControl;
+        private long _expireAt;
+
+        Tracker(String id, RateControl rateControl)
+        {
+            _id = id;
+            _rateControl = rateControl;
+            _expireAt = NanoTime.now() + TimeUnit.SECONDS.toNanos(2);
```

Comment:
Hardcoded parameter `2` should be made configurable, as idleness of a tracker is subjective: for a connection that usually get 10k requests/s, 2 seconds is a long time, but not for a connection that usually gets 1 request/s.